src/rust/engine/client/src/pantsd.rs

```diff
@@ -0,0 +1,247 @@
+    format!(
+      "\
+      The last pid for the pantsd controlling {working_dir} was {pid} but it's no longer running: \
+      {err}\
+      ",
+      working_dir = working_dir.display(),
+      pid = pid,
+      err = e
+    )
+  })?;
+
+  // Check that the live process is in fact the expected pantsd process (i.e.: pids have not
+  // wrapped).
+  let pantsd_process = remoteprocess::Process::new(pid).map_err(|e| {
+    format!(
+      "Failed to read process information for pantsd at pid {pid}: {err}",
+      pid = pid,
+      err = e
+    )
+  })?;
+  let expected_process_name_prefix = pantsd_metadata.process_name()?;
+  let actual_command_line = pantsd_process.cmdline().map_err(|e| {
+    format!(
+      "Failed to determine the process name for the running process at pid {pid}: {err}",
+      pid = pid,
+      err = e
+    )
+  })?;
+  let actual_argv0 = actual_command_line.get(0).ok_or_else(|| {
+    format!(
+      "The command line for pantsd at pid {pid} was unexpectedly empty.",
+      pid = pid
+    )
+  })?;
+  // It appears the the daemon only records a prefix of the process name, so we just check that.
+  if actual_argv0.starts_with(&expected_process_name_prefix) {
+    Ok(port)
+  } else {
+    Err(format!(
+      "\
+      The process with pid {pid} is not pantsd. Expected a process name matching {expected_name} \
+      but is {actual_name}.\
+      ",
+      pid = pid,
+      expected_name = expected_process_name_prefix,
+      actual_name = actual_argv0
+    ))
+  }
+}
+
+#[cfg(test)]
+mod test {
+  use crate::build_root::BuildRoot;
+  use crate::pantsd;
+  use std::fs;
+  use std::net::TcpStream;
+  use std::process::{Command, Stdio};
+  use std::str::from_utf8;
+  use tempdir::TempDir;
+
+  fn launch_pantsd() -> (BuildRoot, TempDir) {
+    let build_root = BuildRoot::find()
+      .expect("Expected test to be run inside the Pants repo but no build root was detected.");
+    let pants_subprocessdir = TempDir::new("pants_subproccessdir").unwrap();
+    let mut cmd = Command::new(build_root.join("pants"));
+    cmd
+      .current_dir(build_root.as_path())
+      .arg("--pants-config-files=[]")
+      .arg("--no-pantsrc")
+      .arg("--pantsd")
+      .arg(format!(
+        "--pants-subprocessdir={}",
+        pants_subprocessdir.path().display()
+      ))
+      .arg("-V")
+      .stderr(Stdio::inherit());
+    let result = cmd.output().unwrap();
+    assert_eq!(Some(0), result.status.code());
+    assert_eq!(
+      fs::read_to_string(
+        build_root
+          .join("src")
+          .join("python")
+          .join("pants")
+          .join("VERSION")
+      )
+      .unwrap(),
+      from_utf8(result.stdout.as_slice()).unwrap()
+    );
+    (build_root, pants_subprocessdir)
+  }
+
+  fn assert_connect(port: u16) {
+    assert!(
+      port >= 1024,
+      "Pantsd should never be running on a privileged port."
+    );
+
+    let stream = TcpStream::connect(("0.0.0.0", port)).unwrap();
```

Comment:
Same question as earlier regarding `0.0.0.0`.

Reply:
I missed the earlier question. I just looked at the server end of this though and we bind explicitly to the loopback so I can switch to that.

Reply:
Fixed.